aiida/repository/backend/abstract.py

```diff
@@ -50,17 +50,25 @@ def erase(self):
         """
 
     @staticmethod
-    def is_readable_byte_stream(handle):
+    def is_readable_byte_stream(handle) -> bool:
         return hasattr(handle, 'read') and hasattr(handle, 'mode') and 'b' in handle.mode
+    
+    def check_byte_stream(self, handle: typing.Any) -> None:
+        """Check that a handle is a valid byte stream.
+        
+        :param handle: filelike object with the byte content.
+        :raises TypeError: if the handle is not a byte stream.
+        """
+        if not isinstance(handle, io.BytesIO) and not self.is_readable_byte_stream(handle):
+            raise TypeError(f'handle does not seem to be a byte stream: {type(handle)}.')
 
-    def put_object_from_filelike(self, handle: io.BufferedIOBase) -> str:
+    @abc.abstractmethod
+    def put_object_from_filelike(self, handle: typing.BinaryIO) -> str:
         """Store the byte contents of a file in the repository.
 
         :param handle: filelike object with the byte content to be stored.
         :return: the generated fully qualified identifier for the object within the repository.
         """
-        if not isinstance(handle, io.BytesIO) and not self.is_readable_byte_stream(handle):
```

Comment:
mypy didn't like this, and it feels like it goes a bit against the concept of an abstract method, so I moved it to a separate method. But its not a deal breaker

Reply:
I guess the best way to do this technically is something like:

```python

    def put_object_from_filelike(self, handle: typing.BinaryIO) -> str:
        """Store the byte contents of a file in the repository.
        :param handle: filelike object with the byte content to be stored.
        :return: the generated fully qualified identifier for the object within the repository.
        """
		check_byte_stream(handle)
        return _put_object_from_filelike(handle)

    @abc.abstractmethod
    def _put_object_from_filelike(self, handle: typing.BinaryIO) -> str:
```

then the concrete methods do not have to remember to call `super()`

Reply:
done